### ISSUE TITLE
Comparison gives stackoverflow.

### ISSUE THREAD

**Issue by PixelOcelot** [decibel/variant#5]
Issue is found for [9b4ca82](https://github.com/decibel/variant/commit/9b4ca82f0560953fd9ce55426c874c7ca601ecff) commit.
I'm using PostgreSQL 9.4 x86. Extension built with msvc.

Reproduce steps:

```

CREATE EXTENSION variant;
SELECT variant.register('vtest', ARRAY['integer'::regtype, 'boolean'::regtype, 'text'::regtype], true);

CREATE TABLE test( v variant.variant(vtest) NOT NULL);

INSERT INTO test VALUES (42);
INSERT INTO test VALUES ('test'::text);

CREATE OPERATOR < (
  PROCEDURE = _variant.variant_lt
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = >
  , NEGATOR = >=
);
CREATE OPERATOR <= (
  PROCEDURE = _variant.variant_le
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = >=
  , NEGATOR = >
);
CREATE OPERATOR = (
  PROCEDURE = _variant.variant_eq
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = =
  , NEGATOR = !=
);
CREATE OPERATOR != (
  PROCEDURE = _variant.variant_ne
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = !=
  , NEGATOR = =
);
CREATE OPERATOR >= (
  PROCEDURE = _variant.variant_ge
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = <=
  , NEGATOR = <
);
CREATE OPERATOR > (
  PROCEDURE = _variant.variant_gt
  , LEFTARG = variant.variant
  , RIGHTARG = variant.variant
  , COMMUTATOR = <
  , NEGATOR = <=
);

SELECT * FROM test WHERE v > 0;
```

Error: 

ERROR:  stack depth limit exceeded
HINT:  Increase the configuration parameter "max_stack_depth" (currently 2048kB), after  ensuring the platform's stack depth limit is adequate.
CONTEXT:  SQL-statement : "SELECT variant_name, variant_enabled, storage_allowed, allowed_types @> array[ $2 ] FROM variant._registered WHERE variant_typmod = $1"
SQL-statement : "SELECT CASE WHEN $1 = $2 THEN 0 WHEN $1 < $2 THEN -1 ELSE 1 END::int"
SQL-statement : "SELECT CASE WHEN $1 = $2 THEN 0 WHEN $1 < $2 THEN -1 ELSE 1 END::int"
SQL-statement : "SELECT CASE WHEN $1 = $2 THEN 0 WHEN $1 < $2 THEN -1 ELSE 1 END::int"
...
and so forth.

**Comment by decibel** [comment]
What command were you running?

**Comment by decibel** [comment]
BTW, does the MSVC stuff work OK or did you have to change anything? I have no way to test or work on that stuff, so pull requests to it very welcome.

**Comment by PixelOcelot** [comment]
Last select is the command which gives errors:

    SELECT * FROM test WHERE v > 0;

I haven't used msvs solution you provided. Because I've started work with version from pgxn, worked out how to build it on windows, then I found github repo and just repeated process and only later I noticed that there was prepared solution :) 

Anyway, I'll test your solution tomorrow and tell you results.

**Comment by decibel** [comment]
Ahh, I missed that in the test case. I'm able to reproduce it here; I'll try to take a look at it later today.

**Comment by decibel** [comment]
I know what's going on; currently variant_cmp_int punts back to Postgres to compare two different data types. The problem is that because of how the casting is setup, if there's no other valid comparison it's going to decide to try and cast both operators to variant and compare that... nice recursion. I need to figure out how to detect that and throw an appropriate error instead.

**Comment by decibel** [comment]
BTW, in the meantime, you can work around this by:

SELECT * FROM test WHERE v::int > 0;
ERROR:  invalid input syntax for integer: "test"
CONTEXT:  SQL statement "SELECT $1::integer"

**Comment by decibel** [comment]
Err, scratch that... it's not a great idea to just randomly cast things.

You could test to see what the original type is using variant.original_type(v), but then you'd need to know what it would be safe to accept, which is not necessarily a simple thing.

BTW, what do you plan on using variant for? There's been discussion in the community about adding this, but it always grinds to a halt when people ask what the practical use case is.

**Comment by PixelOcelot** [comment]
Maybe it would be better to do this check internaly in compare functions? And if operands types are different then throw warning about implicit cast or error demanding explicit cast?

I'm using Postgre as SQL hub via foreign data wrappers. Data sources operate with variant data type. 
There're a lot of historical data and it's possible to have different data types in the result of the single query.
Have to support it :)

**Comment by decibel** [comment]
> Maybe it would be better to do this check internaly in compare
> functions? And if operands types are different then throw warning about
> implicit cast or error demanding explicit cast?

Yeah, I need to manually determine what operator to use and deal with it 
when the coercion logic comes back with "here, cast to variant and then 
compare". Unfortunately it's a non-trivial change.

> I'm using Postgre as SQL hub via foreign data wrappers. Data sources
> operate with variant data type.
> There're a lot of historical data and it's possible to have different
> data types in the result of the single query.
> Have to support it :)

So the other databases are using variants?

**Comment by PixelOcelot** [comment]
>  So the other databases are using variants?
>

Yes. And not only databases. My data sources are not relational.

Here are a bunch of references:
https://msdn.microsoft.com/en-us/library/office/gg251448.aspx - VB reference for office developers.
https://msdn.microsoft.com/ru-ru/library/ms173829.aspx - MS SQL use it for processing and storing data.
https://msdn.microsoft.com/ru-ru/library/windows/desktop/ms714931(v=vs.85).aspx - same you can find in OLE DB reference.
http://docs.oracle.com/cd/B19306_01/appdev.102/b14258/t_anydat.htm#BEHEICHI - Oracle has it too.

**Comment by PixelOcelot** [comment]
>Unfortunately it's a non-trivial change.
>

Check out https://technet.microsoft.com/en-us/library/ms181071(v=sql.105).aspx  "sql_variant Comparisons" section.

Also another solution: compare function can respond with not only "greater", "lesser", "equal" results but with "impossible to compare" too. Later can be leade to default result: one operand will be decided to be lesser than other according to some convention (microsoft solved it with hierarchy of data types, see reference above, here it can be done simplier) or error/messages.

**Comment by decibel** [comment]
Could you elaborate on how variants are being used in the MSSQL databases? I know that other databases support the feature, but that hasn't been a strong enough argument to sway the community.

As for the fix, for now I'll probably just throw an exception; that's not the hard part. The hard part is re-implementing all the operator logic that the database already has. I'll be able to copy some of it out of the array/record handling code, but probably not all of it. I've got some client work going on right now, so I'm not sure how long that'll take me. Unless you want to take a crack at that?

**Comment by decibel** [comment]
I've done some work on this in https://github.com/decibel/variant/tree/operators.

Right now I've got code that hopefully knows how to correctly find a new arbitrary operator that will NOT cycle back to variant_op(). Right now it's building an expression node, but in retrospect if that's not good enough for RI triggers then it's doubtful it's good enough for us. Basically, I think we need to duplicate ri_GenerateQual() and friends and use that to create a query we can then hand to SPI. This is similar to what the existing code does, except:

- This should handle any 2 argument operator, not just comparators
- It would use the OPERATOR() construct to ensure that we're getting a specific operator

**Comment by PixelOcelot** [comment]
>Unless you want to take a crack at that?
>
Sorry, have to work on different project right now. Don't have time for that now. But I'll definetly return and probably will help you with this problem and documentation too.
Can't say much on your solution right now. Only that CI build wasn't able to digest it :)

**Comment by decibel** [comment]
Yes, that branch is definitely not in a working state right now. :)

**Comment by decibel** [comment]
I've done more work on this. I now have a general purpose C function that can execute any operator that references it, but only if the calling SQL doesn't get inlined. So your test case works because it involves a table, but something like SELECT 1::int::variant.variant < 2 blows up because the operator gets replaced with a direct function call, which means the function can no longer determine what the original caller was. I don't see any good way to fix that. :(
The last possible trick I can think of is to try using a SQL function for each individual operator and have that function pass in what the original operator was.
